thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+both in Go.
+
+Before I get into the experience, let me preface this by mentioning that Rust and Go don't exactly target the same audiences.
+Go is garbage collected and is okay with losing out on some performance for ergonomics;
+whereas Rust tries to keep everything as a compile time check as much as possible.
+This usually makes Rust much more useful for lower level applications.
+
+In my specific situation, however, I was playing around with distributed systems via threads
+(or goroutines), so this fit perfectly into the area of applicability of both languages.
+
+
+This article isn't exactly intended to be a comparison between the two. I understand that as a newbie at Go,
+I'll be trying to do things the wrong way and make bad conclusions off of this. My way of coding may not be the
+"Go way" (I'm mostly carrying over my Rust style to my Go code since I don't know better); so everything may seem
+like a hack to me. Please keep this in mind whilst reading the post, and feel free to let me know the "Go way"
+of doing the things I was stumbling with.
+
+This is more of a sketch of my experiences with the language, specifically from the point of
+view of someone coming from Rust; used to the Rusty way of doing things. It might be useful as an
+advisory to Rustaceans thinking about trying the language out, and what to expect.
+
+[raft]: https://github.com/Manishearth/cs733/tree/master/assignment3/raft
+[dash]: http://manishearth.github.io/dash
+
+// I've structured this as "what I liked"/"what I didn't like", but I'm open to other ways of structuring this too
+
+## What I liked
+
+Despite the performance costs, having a GC at your disposal after using Rust for very long is quite liberating. 
+For a while my internalized borrow checker would throw red flags on me tossing around data indiscriminately,
+but I learned to  ignore it as far as Go code goes. I was able to quickly share state via pointers without
+worrying about safety, which was quite useful.
+
+Having channels as part of the language itself was also quite ergonomic.  The `data <- chan` and `chan <- data` syntax is fun to use,
+and whilst it's not very different from `.send()` and `.recv()` in Rust, I found it surprisingly easy to read.
+Initially I got confused often by which side the channel was, but after a while I got used to it.
+It also has an in built select block for selecting over channels (Rust has a macro for the same purpose, but no language support).
+
+
+`gofmt`. The Go style of coding is different from the Rust one (tabs vs spaces, how declarations look),
+but I continued to use the Rust style because of the muscle memory (also too lazy to change the settings
+in my editor). `gofmt` made life easy since I could just run it in a directory and it would fix everything.
+Eventually I was able to learn the proper style by watching my code get corrected.
+I'd love to see a `rustfmt`, however!
```

Comment:
I've recently started using the similar tool `standard` for my JavaScript libraries:

https://github.com/feross/standard

and love it.